[ot_certs] ASN1 codegen

pamaury opened the pull request:
Only review the last three commits, others are from #20601. See commits for description.
To test:
```bash
# build all files
./bazelisk.sh build //sw/device/silicon_creator/lib/cert/...
# view all targets:
./bazelisk.sh query //sw/device/silicon_creator/lib/cert/...
# query list of output for a certificate:
./bazelisk.sh outquery-all //sw/device/silicon_creator/lib/cert:example_template
# use your favorite editor to look at those files
```

jwnrt commented on sw/host/ot_certs/src/codegen.rs:
```diff
@@ -0,0 +1,408 @@
+
+status_t execute_test(ujson_t uj, FILE *outf) {
+  values_t values;
+  TRY(ujson_deserialize_values_t(&uj, &values));
+  size_t cert_size = 1000;
+  uint8_t *cert = malloc(cert_size);
+  TRY(build_cert(&values, cert, &cert_size));
+  fwrite(cert, cert_size, 1, outf);
+  fclose(outf);
+  return OK_STATUS();
+}
+
+int main(int argc, char **argv) {
+  if (argc != 3) {
+    fprintf(stderr, "usage: %s <values.json> <output.pem>\n", argv[0]);
+    return 1;
+  }
+  FILE *inf = fopen(argv[1], "r");
+  if (inf == NULL) {
+    perror("cannot open json file with values");
+    return 2;
+  }
+  FILE *outf = fopen(argv[2], "wb");
+  if (outf == NULL) {
+    perror("cannot open output certificate file with values");
+    return 2;
+  }
+  ujson_t ujson = ujson_init(inf, fgetc_wrapper, putbuf_wrapper);
+  status_t err = execute_test(ujson, outf);
+  const char *code = NULL;
+  int32_t arg = 0;
+  char mod_id[5];
+  bool is_err = status_extract(err, &code, &arg, mod_id);
+  mod_id[4] = 0;
+  if(is_err) {
+    fprintf(stderr, "An error occured: %s[%d,%s]\n", code, arg, mod_id);
+  }
+  return 0;
+}
+"#;
```
Nit / suggestion: you may find it clearer to put this in a `.c` file in this directory and use `include_str!` to pull it in.

jwnrt commented on rules/certificates.bzl:
```diff
@@ -0,0 +1,75 @@
+        ),
+    ]
+
+# TODO: document
```
Would you mind adding these doc strings before we merge?

It could easily be forgotten if we leave it to the future

jwnrt reviewed (approved):
I've done a pass on this and it looks generally okay, but there's really lots of generation code and I definitely haven't checked its correctness in depth.

I don't have any refactorings to suggest, but if you have anything that could make this easier to follow or to make it more compact, that could be helpful for the future.

pamaury commented:
Thanks for the comments James, I will rework that. Regarding the code generation, I agree that it's a lot of code but I don't have any great idea to make the code more compact. I'll think about it when updating the code.

pamaury commented:
I have removed the code of the writer from this PR since it takes quite a bit of code and it is not essential. I have addressed some changes, I still need to fix two TODO.

timothytrippel commented on sw/device/silicon_creator/lib/cert/BUILD:
```diff
@@ -2,6 +2,36 @@
+)
+
+certificate_template(
+    "generic_template",
+    "generic.hjson",
```
```suggestion
    name = "generic_template",
    template = "generic.hjson",
```

applies throughout

timothytrippel commented on rules/certificates.bzl:
```diff
@@ -0,0 +1,88 @@
+# This rule uses `opentitantool certificate codegen` to generate the
+# source files of the C certificate generator. Those source files
+# need to be compiled and linked with the asn1 library: this is NOT
+# done by this rule. The generates files are put in the following
```
```suggestion
# done by this rule. The generated files are put in the following
```

timothytrippel commented on rules/certificates.bzl:
```diff
@@ -0,0 +1,88 @@
+            "--output-c={}".format(out_c.path),
+            "--output-h={}".format(out_h.path),
+        ],
+        executable = tc.tools.opentitantool,
```
we probably also want to hit the generated C files with clang format after they are generated?

pamaury replied:
I will look into it.

jwnrt commented on sw/host/ot_certs/src/asn1/codegen.rs:
```diff
@@ -0,0 +1,515 @@
+                // NOTE: we are not using a constant for the size of the string because technically there is no way
+                // to know what the `char` type is and therefore what is the length of this string for the C compiler.
+                self.push_str_with_indent(&format!("RETURN_IF_ERROR(asn1_push_string(&state, {str_type}, \"{x}\", strlen(\"{x}\")));\n"));
+                // A tagged string needs a tag, up to 3 bytes of length and the string itself.
```
Nit: can we put `(up to three bytes)` in parens?

jwnrt commented on sw/host/ot_certs/src/asn1/codegen.rs:
```diff
@@ -0,0 +1,515 @@
+                self.push_str_with_indent(&format!("RETURN_IF_ERROR(asn1_push_string(&state, {str_type}, \"{x}\", strlen(\"{x}\")));\n"));
+                // A tagged string needs a tag, up to 3 bytes of length and the string itself.
+                // Don't try to exactly compute how many bytes we need for the length.
+                self.max_out_size += 1 + 3 + x.len();
```
Nit: if the `1` is for the null terminator, can we put it at the end to make that clear?

jwnrt commented on sw/host/ot_certs/src/codegen.rs:
```diff
@@ -0,0 +1,287 @@
+
+        // This file was automatically generated using opentitantool from:
+        // {from_file}
+        "#};
```
Nit: de-indent this one level

jwnrt reviewed (approved):
Reviewed this with @luismarques and we only had these nits

pamaury commented:
I have pushed an update that rebases the code and also addresses the various comments. I have also improved the documentation and the code now produces better estimates for the size of the certificate (previous estimates were in the ball park of 550 bytes, now it is at 440 so it's much better!).

nbdd0121 commented on sw/host/ot_certs/src/codegen.rs:
```diff
@@ -0,0 +1,318 @@
+    source_c.push_str(&license_and_warning);
+    source_h.push_str(&license_and_warning);
+    let preproc_guard_include = tmpl.name.to_uppercase();
+    source_h.push_str(&format!("#ifndef __{}__\n", preproc_guard_include));
```
```suggestion
    write!(source_h, "#ifndef __{preproc_guard_include}__\n");
```
